quinn/src/streams.rs

```diff
@@ -149,6 +151,27 @@ where
         conn.wake();
     }
 
+    /// Completes if/when the peer stops the stream, yielding the error code
+    pub fn stopped(&mut self) -> Stopped<'_, S> {
+        Stopped { stream: self }
+    }
+
+    #[doc(hidden)]
+    pub fn poll_stopped(&mut self, cx: &mut Context) -> Poll<VarInt> {
+        let mut conn = self.conn.lock().unwrap();
+        if self.stopped_event.is_none() {
+            let (send, recv) = oneshot::channel();
+            self.stopped_event = Some(recv);
+            conn.stopped.insert(self.stream, send);
+        }
+
+        self.stopped_event
+            .as_mut()
+            .unwrap()
+            .poll_unpin(cx)
+            .map(|x| x.unwrap())
```

Comment:
Should we return an error rather than hanging if the connection is lost? Hung futures strike me as a footgun in general, since they produce leaks if you block a task on them. It's hard to imagine why anyone would block a task solely on waiting for a stream to become stopped, since that might never happen and the `&mut` prevents you from doing anything more useful with the stream in the mean time, so I'm not sure, but my inclination is to favor consistency with the other futures (which all yield errors) and minimize potential for surprising behavior.

This should be as simple as checking `conn.error` when polling the oneshot yields `Pending`, as we're doing in `poll_finish`.